numba/cuda/simulator/vector_types.py

```diff
@@ -0,0 +1,60 @@
+from numba import types
+from numba.cuda.stubs import _vector_type_stubs
+
+
+class SimulatedVectorType:
+    attributes = ['x', 'y', 'z', 'w']
+
+    def __init__(self, *args):
+        args_flattened = []
+        for arg in args:
+            if isinstance(arg, SimulatedVectorType):
+                args_flattened += arg.as_list()
+            else:
+                args_flattened.append(arg)
+        self._attrs = self.attributes[:len(args_flattened)]
+        if not self.num_elements == len(args_flattened):
+            raise TypeError(
+                f"{self.name} expects {self.num_elements}"
+                f" elements, got {len(args_flattened)}"
+            )
+        assert self.num_elements == len(args_flattened), \
+            f"{self.num_elements} != {len(args_flattened)}"
+
+        for arg, attr in zip(args_flattened, self._attrs):
+            setattr(self, attr, arg)
+
+    @property
+    def name(self):
```

Comment:
I think this and `num_elements` get replaced when a new type is made from `SimulatedVectorType` in `make_simulated_vector_type()` below, and that these methods are only placeholders. Is this correct?

Reply:
That's correct, they are replaced at https://github.com/isVoid/numba/blob/4faf3f58b3192212b5ae91eff07cedd2f0ae3b2a/numba/cuda/simulator/vector_types.py#L41-L43. On a second note they seem unrelated though.